src/gxcore/sd/calendar/schedule.ts

```diff
@@ -0,0 +1,127 @@
+import { notImplemented } from "../../../misc/helpers";
+import { addHours } from "../../../datetime/addHours";
+import { GUID } from "../../../types/guid";
+
+/**
+ * Allows scheduling some task on the end user's calendar. Every parameter is optional except for `title` and `startDate`.
+ * @param {string} title
+ * @param {Date} startDate
+ * @param {Date} endDate
+ * @param {Date} startTime
+ * @param {Date} endTime
+ * @param {string} place
+ */
+export const schedule = (
+  title: string,
+  startDate: Date,
+  endDate: Date,
+  startTime: Date,
+  endTime: Date,
+  place: string
+) => {
+  let sDateTime = getDateTime(startDate, startTime);
+  let eDateTime = getDateTime(endDate, endTime, addHours(sDateTime, 1));
+
+  let calStr = createCalendarEvent(sDateTime, eDateTime, title, place);
+
+  downloadCalendar(calStr);
+};
+
+function convertDateTimeToString(dt: Date): string {
+  // 20190814T160000Z
+  return (
+    dt.getUTCFullYear().toString() +
+    stringFromTimeNumber(dt.getUTCMonth() + 1) +
+    stringFromTimeNumber(dt.getUTCDate()) +
+    "T" +
+    stringFromTimeNumber(dt.getUTCHours()) +
+    stringFromTimeNumber(dt.getUTCMinutes()) +
+    stringFromTimeNumber(dt.getUTCSeconds()) +
+    "Z"
+  );
+}
+
+function stringFromTimeNumber(num: number): string {
+  return (num < 10 ? "0" : "") + num.toString();
+}
+
+function getDateTime(
+  datePart: Date,
+  timePart: Date,
+  defaultValue?: Date
+): Date {
+  if (!datePart) {
+    return defaultValue;
+  }
+
+  let result = datePart;
+  if (timePart) {
+    result.setHours(timePart.getHours());
+    result.setMinutes(timePart.getMinutes());
+    result.setSeconds(timePart.getSeconds());
+  }
+  return result;
+}
+
+export function createCalendarEvent(
+  start: Date,
+  end: Date,
+  title: string,
+  place: string
+): string {
+  return (
```

Comment:
I think you could take advantage of [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) here